compiler/ilgen/IlBuilder.cpp

```diff
@@ -1565,6 +1565,119 @@ IlBuilder::AtomicAdd(TR::IlValue * baseAddress, TR::IlValue * value)
+ *    +---------------------------------+                         |                                  |
+ *                       |                                        |                                  |
+ *                       v                                        V                                  V
+ *    +-----------------------------------------+   +-------------------------------+    +-------------------------+     
+ *    |<block_$fallThrough>                     |   |<block_$persistentFailure>     |    |<block_$transientFailure>|      
+ *    |====================                     |   |===========================    |    |=========================|
+ *    |     add (For illustration, we take add  |   |AtomicAdd (For illustration, we|    |   ...                   |
+ *    |     ... as an example. User could       |   |...       take atomicAdd as an |    |   ...                   |
+ *    |     ... replace it with other non-atomic|   |...       example. User could  |    |   ...                   |
+ *    |     ... operations.)                    |   |...       replace it with other|    |   ...                   |
+ *    |     ...                                 |   |...       atomic operations.)  |    |   ...                   |
+ *    |     TEND                                |   |...                            |    |   ...                   |
+ *    |goto --> block_$merge                    |   |goto->block_$merge             |    |goto->block_$merge       |
+ *    +----------------------------------------+    +-------------------------------+    +-------------------------+
+ *                      |                                          |                                 |
+ *                      |                                          |                                 |
+ *                      |------------------------------------------+---------------------------------+
+ *                      |
+ *                      v             
+ *              +----------------+     
+ *              | block_$merge   |     
+ *              | ============== |   
+ *              |      ...       |
+ *              +----------------+   
+ *
+ * \endverbatim 
+ *
+ * \structure & \param value
+ *     tstart
+ *       |
+ *       ----persistentFailure // This is a permanent failure, try atomic way to do it instead
+ *       |
+ *       ----transientFailre // Temporary failure, user can retry 
+ *       |
+ *       ----fallthrough // Success, use general(non-atomic) way to do it
+ */
+void
+IlBuilder::TransactionBegin(TR::IlBuilder **persistentFailureBuilder, TR::IlBuilder **transientFailureBuilder, TR::IlBuilder **fallThroughBuilder)
+    {   
+    //ILB_REPLAY("%s->TransactionBegin(%s, %s, %s);", REPLAY_BUILDER(this), REPLAY_BUILDER(persistentFailureBuilder), REPLAY_BUILDER(transientFailureBuilder), REPLAY_BUILDER(fallThroughBuilder)); 
```

Comment:
you should do the createBuilderIfNeeded calls before you do the ILB_REPLAY line.